### ISSUE TITLE
radiff2 segmentation fault for function-level diffing

### ISSUE THREAD

**Issue by computereasy** [radare/radare2#5386]
Hello, I would like to use radiff2 to launch some function-level diffing towards 64-bit x86 binary code. Basically the binary code is all `unstrip`, meaning it has the function information, I just want to compare the similarity between two functions. 

However, I got some segmentation fault as below when testing binaries (note that binary below are `strip`, but similar errors happen on `unstrip` binary code):

    `☁  coreutils [master] ⚡ radiff2 -A -C /bin/true /bin/false                               
    [x] Analyze all flags starting with sym. and entry0 (aa)
    [x] Analyze len bytes of instructions for references (aar)
    [x] Analyze function calls (aac)
    [ ] [*] Use -AA or aaaa to perform additional experimental analysis.
    [x] Constructing a function name for fcn.* and sym.func.* functions (aan))
    [x] Analyze all flags starting with sym. and entry0 (aa)
    [x] Analyze len bytes of instructions for references (aar)
    [x] Analyze function calls (aac)
    [ ] [*] Use -AA or aaaa to perform additional experimental analysis.
    [x] Constructing a function name for fcn.* and sym.func.* functions (aan))
    *** glibc detected *** radiff2: munmap_chunk(): invalid pointer: 0x00007f03985c54a0 ***
    [1]    27928 segmentation fault (core dumped)  radiff2 -A -C /bin/true /bin/false
`

**Comment by computereasy** [comment]
Another test on unstripped binary code:

    `☁  coreutils [master] ⚡ radiff2 -AC ./icc_coreutils_64_O0_base64 ./icc_coreutils_64_O1_base64
    [x] Analyze all flags starting with sym. and entry0 (aa)
    [x] Analyze len bytes of instructions for references (aar)
    [x] Analyze function calls (aac)
    [ ] [*] Use -AA or aaaa to perform additional experimental analysis.
    [x] Constructing a function name for fcn.* and sym.func.* functions (aan))
    [x] Analyze all flags starting with sym. and entry0 (aa)
    [x] Analyze len bytes of instructions for references (aar)
    [x] Analyze function calls (aac)
    [ ] [*] Use -AA or aaaa to perform additional experimental analysis.
    [x] Constructing a function name for fcn.* and sym.func.* functions (aan))
    *** glibc detected *** radiff2: free(): invalid next size (fast): 0x00007f80e7374530 ***
    ======= Backtrace: =========
    /lib/x86_64-linux-gnu/libc.so.6(+0x7da26)[0x7f80e4358a26]
    /usr/lib/libr_util.so(r_diff_buffers_distance+0x583)[0x7f80e46e65ad]
    /usr/lib/libr_anal.so(r_anal_diff_bb+0x177)[0x7f80e54c696b]
    /usr/lib/libr_anal.so(r_anal_diff_fcn+0x31d)[0x7f80e54c6e88]
   /usr/lib/libr_core.so(r_core_gdiff+0x2a3)[0x7f80e621da92]
    radiff2(main+0x62d)[0x7f80e66b7aa6]
    /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xed)[0x7f80e42fc7ed]
    radiff2(+0x1519)[0x7f80e66b6519]
    ======= Memory map: ========
    7f80e15cf000-7f80e15e4000 r-xp 00000000 fc:02 13107244                   /lib/x86_64-linux-gnu/libgcc_s.so.1
    (a long map of memory...)
`

**Comment by radare** [comment]
share the bins or use valgrind to give better understanding of the crash . also confirm that you are using latest r2 from git

**Comment by computereasy** [comment]
@radare I am using the latest version. Let me share the bins.

**Comment by computereasy** [comment]
@radare  The first two binaries (true & false):
https://www.dropbox.com/s/ywr1ryjrqh9n9m3/false?dl=0
https://www.dropbox.com/s/kefpefugqq9jmod/true?dl=0

**Comment by computereasy** [comment]
@radare The second two binaries. Note that they are unstripped and they have the function symbols:
https://www.dropbox.com/s/s0jt1vlcy9jf10o/icc_coreutils_64_O0_base64?dl=0
https://www.dropbox.com/s/qho4v7ocld05tae/icc_coreutils_64_O1_base64?dl=0

**Comment by Maijin** [comment]
Paste output of `r2 -v` here

**Comment by computereasy** [comment]
@Maijin 

 radare2 0.10.5-git 11921 @ linux-x86-64 git.0.10.4-82-ge62aac4
commit: e62aac49f3a759a45342e6ae4b51a139921a24d7 build: 2016-07-26

**Comment by radare** [comment]
omg the last pr optimizing this code introduced a huge amount of buffer overflows and underflows. i have a partial fix but i want to check this with care before pushing.

can those bins be shared? a test clase would be great to avoid those things to happen again and ensure radiff2 works as expected

**Comment by Maijin** [comment]
@computereasy see: https://github.com/radare/radare2-regressions/blob/master/t.tools/radiff2/radiff2

**Comment by computereasy** [comment]
@radare  Yes, those are all released bin.

**Comment by computereasy** [comment]
@Maijin  sorry, what am I expected to do..?

**Comment by Maijin** [comment]
```
can those bins be shared? a test clase would be great to avoid those things to happen again and ensure radiff2 works as expected
```

Just look at the example to create a test 👍

**Comment by computereasy** [comment]
@Maijin  I got it. But without a workable `radiff2`, I have no idea what the EXPECT should be....

**Comment by Maijin** [comment]
Just output to dev/null and check for no result, this should be sufficient to check if segfault or not and mark it as broken=1

**Comment by alvarofe** [comment]
I fixed the crashes though the code must be review with more care i failed when doing the review at the moment.

**Comment by radare** [comment]
maybe contact the author as long as he knew the technical details of the algorithm for proper review and write more tests after this.

**Comment by zonkzonk** [comment]
do folks still need a valgrind output?

**Comment by radare** [comment]
Nope

**Comment by Maijin** [comment]
https://github.com/radare/radare2/pull/5316 @NikolaiHampton

**Comment by NikolaiHampton** [comment]
Sorry guys, I have zero time right now. Might be better for you to revert if it's causing issues. 

I have processed about 5000 unique files (10,000 pairs) through radiff2 and haven't had any segfaults or strange terminations. I believe the crashes are real, I'm just saying I haven't seen them, so I'll need to give it some time to investigate.

Unfortunately due to workload, this won't happen for another week or two. 

Sorry guys.

**Comment by computereasy** [comment]
@NikolaiHampton  Hello, could you please shed some light on my following questions?  I just don't understand what is meaning of "1.000000" regarding "UNMATCH". #5388   I understand that you are super busy, but this is very critical to me... Thank you!

**Comment by NikolaiHampton** [comment]
Hey @computereasy, I don't know where the UNMATCH is coming from, I assume it's whatever is calling the r_diff_buffers_distance().. 1.000000 should mean it's identical, but there could be an issue with the comparison operation in whatever is calling r_diff_buffers_distance.. The similarity value returned is a double, if that helps. I'd love to have a look but I'm very far behind schedule on paper I'm finishing. When do you need this by, I'll see what I can squeeze in!

**Comment by computereasy** [comment]
@NikolaiHampton  No problem. I can only rely on this numeric value and ignore the "MATCH/UNMATCH" for now. Good luck with your paper!

**Comment by NikolaiHampton** [comment]
@radare @alvarofe - My own test cases for calculation accuracy were based on only very small files as the original r_diff_buffers_distance couldn't handle large files (due to memory and CPU requirements - originally I couldn't run anything larger than ~40KB).

My test case included a few dozen files created at random and compared to see that the results with pre #5000 runs of: radiff2 -s A B were the same.

As far as segfaults or memory issues go, using a large file, my valgrind results were the same with r_diff_buffers_distance content commented out and simply running:

```
r_diff_buffers_distance(...) {
    *distance=0;
    *similarity=1;
    return true;
}
```
This valgrind showed a leak! So yes it's definitely leaking somewhere, but I'm not sure where that could be.

I think @alvarofe found an access of one byte past the end of the buffer which I thought was due to a comparison on FileA[i+1] (when i is equal to the size of A).. But I haven't checked again.

Has someone got a stack trace of a current segfault and the input files that caused the issue? 

Could the segfaults be happening elsewhere? Is there another buffers_distance() util that's being called? Really not trying to pass the buck, just throwing ideas around while I don't have time to investigate.

**Comment by NikolaiHampton** [comment]
@computereasy I too am using radiff for malware analysis and router firmware comparisons! :)

Be careful with relying on the 1.00000 for comparison.

try running:
radiff2 -v fileA fileB

You'll see there's a distance (the number of bytes changed, added or removed), and the similarity.

You might be safe with declaring 1.00000 a match (I suspect you will be, but this isn't ideal). But, I guess it's possible on a very large function of having a distance of say 1, which is below the significance threshold of 1.00(..whatever..)00 which is the reason for your MATCH/UNMATCH. Seems unlikely to me though, but I don't know.

**Comment by NikolaiHampton** [comment]
@radare @alvarofe - Also, I can't reproduce the segfaults that have been discussed, if someone shoots me a couple of files that are misbehaving, I might be able to have a quick look.

**Comment by Maijin** [comment]
Binaries are already in the issue:

https://github.com/radare/radare2/issues/5386#issuecomment-235355504
https://github.com/radare/radare2/issues/5386#issuecomment-235356315

**Comment by NikolaiHampton** [comment]
@Maijin right you are.. thanks.

**Comment by NikolaiHampton** [comment]
@computereasy I can see from your stack trace that there is definitely crash, but I can't replicate it! :( your sample files run fine in with radiff2 -A -C compiled from the master about an hour ago.

The error suggests an errant free() ?? Perhaps I'm being over zealous with the original free?
```
	int *v0 = (int*) calloc ((bLen + 3), sizeof (int));
	if (!v0) {
		eprintf("Error: cannot allocate %i bytes.", bLen + 3);
		return false;
	}

	int *v1 = (int*) calloc ((bLen + 3), sizeof (int));
	if (!v1) {
		eprintf("Error: cannot allocate %i bytes", 2 * (bLen + 3));
		free(v0);
		return false;
	}
```

The only other free()s I have are at the end of execution..

You have me stumped, sorry guys again no idea why this should be failing, I must be missing something really simple!

**Comment by Maijin** [comment]
Try to build with asan (./sys/asan.sh)

**Comment by NikolaiHampton** [comment]
@radare @Maijin @computereasy :

Ok.. I've built with asan.sh, built with -fsanatize=address in both GCC and clang on MacOS and on an Ubuntu 14.04 Linux. I get no errors. I've tested a few dozen of my files, I've tested all four files provided by @computereasy - I get no errors.

I ran radiff2 it through valgrind on Ubuntu, 0 errors, 0 leaks - for both:
radiff2 -s and radiff2 -A -C using the same four files.

For some reason valgrind on MacOS reports leaks and errors, but if I comment out the entire r_diff_buffers_distance (and simple return true) I still get the same leaks and errors! I have no idea why the errors would be there on MacOS, but the fact that the errors are present even when the whole function is commented out makes me wonder if it's not an artifact of MacOS's.

I really don't know where else to look. Sorry guys, obviously there's something that's beyond my grasp. I didn't think the changes to the code were that complicated or difficult.

The results being produced by radiff2 (including the off-by-one-read fix by @alvarofe) from the latest checkout are exactly the same as pre-changes radiff2.

In summary, this is working with the provided files and the provided commands on two platforms using two different compilers. So, I'm totally out of ideas with regard to segfaults..

**Comment by NikolaiHampton** [comment]
@computereasy - I'm still wary of using the 1.0000 instead of the MATCH and UNMATCH.. In my testing I never saw anything other than 1.000 and 0.000 in the similarity column which concerns me, I'd have thought you'd see at least a few .9998's or something.. I haven't had time to look in to the MATCH and UNMATCH tagging, but there's a few conditions that it's based on, and DIFF seems to only be one of them.

**Comment by NikolaiHampton** [comment]
This just in: I have managed to get radiff2 to segfault once.. The weird thing is it happened after a clean checkout on a VM that I'd snapshotted a few days back.. BUT, the segfault was weird and the core dump stacktrace didn't tell me where the error occurred.

I did a make clean&&make&&sudo make install, without changing any CFLAGS or code (I changed nothing at all), and the problem went away I can't get a second segfault out of it no matter how hard I try. I wonder if it's something to do with old libs that were on my previous system?

**Comment by alvarofe** [comment]
The segfaults that were discussed here were fixed. We should add this in the next release to fuzz since i am not quite familiar with the algorithm and I would fail to catch the issues

**Comment by radare** [comment]
can we close this issue now?

**Comment by Maijin** [comment]
Well
```
@NikolaiHampton Hello, could you please shed some light on my following questions? I just don't understand what is meaning of "1.000000" regarding "UNMATCH". #5388 I understand that you are super busy, but this is very critical to me... Thank you!
```

Is still not resolve so no.

**Comment by radare** [comment]
@alvarofe ^

**Comment by NikolaiHampton** [comment]
@radare @Maijin - The issue raised by @computereasy was the segfault was and that was fixed by @alvarofe.
The August 2 comment regarding "shedding light on how MATCH/UNMATCH" works I think is unrelated to the segfault. I think this should be opened as a different issue.

**Comment by radare** [comment]
Should be fixed, please confirm

**Comment by Maijin** [comment]
Not fixed. Here is the false and true comparison:

![image](https://cloud.githubusercontent.com/assets/1408600/18750565/f4542638-80db-11e6-8d8f-e7198ff4e0cd.png)


Still this `UNMATCH  (1.000000)`

**Comment by radare** [comment]
This can be closed now
